### PR TITLE
Device handling fixes in ORTModule

### PR DESCRIPTION
**Pull request by mrry**
**Description**: Multiple fixes to device handling logic in `ORTModule`.
* The `args` and `kwargs` should be expanded in the call to `super(...).to()`.
* When using multiple devices, the `device_id` should be set on the output `IOBinding`.
* When using multiple devices, the `device_id` should be set on the `CUDAExecutionProvider` in each `InferenceSession`.

**Motivation and Context**
- This fixes an observed `TypeError` where a `tuple, dict` are the observed argument types in the superclass method. Thanks to @harshithapv for uncovering the bug!
- It enables distributed data parallel training on a single multi-GPU machine.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -40,7 +42,9 @@ def _get_device_index(device):
                               inputs[idx].data_ptr())
 
     for value_info in model.graph.output:
-        io_binding.bind_output(value_info.name, device if device.find(':') == -1 else device[:device.find(':')])
+        io_binding.bind_output(value_info.name, device if device.find(':') == -1 else device[:device.find(':')],
```
Comment (kit1980) [inline orttraining/orttraining/python/training/ortmodule.py:45]: Maybe don't convert `self._device` to string when call `_create_iobinding`?
Then you can check `device.type` here instead of string search for ':'

Comment (mrry) [reply orttraining/orttraining/python/training/ortmodule.py:45]: I'm trying to minimize the amount of refactoring in this change, but I do agree that the device handling seems like it could be simplified. I'll let @thiagocrepaldi take that on in a follow up, since he's best placed to know if there are any subtleties here.

**Review by kit1980 — APPROVED** [review]

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]
@mrry Could you add the samples you used to validate both DDP and Zero 1/2?

Yu can copy & paste the samples in new files if it makes easier. I will work with Baiju to get a proper CI workflow going this week

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed_config.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      "params": {
+        "warmup_min_lr": 0,
+        "warmup_max_lr": 0.001,
+        "warmup_num_steps": 1000
+      }
+    },
+    "wall_clock_breakdown": true,
+    "fp16": {
+      "enabled": false
+    }
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed_config.json:21]: missing comma here?
```suggestion
    },
```

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed_config.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    "zero_optimization": {
+        "stage":1,
+        "reduce_bucket_size": 500000000
+    },
+    "zero_allow_untested_optimizer": true
+  }
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed_config.json:27]:
```suggestion
  }

```

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed_config.json (hunk 2 above)
Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed_config.json]: Thanks, this seems to be a stale broken copy of the file... I just pushed again with the fixes.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 4]
```diff
@@ -188,6 +192,13 @@ def forward(self, *inputs, **kwargs):
             self._forward_session = onnxruntime.InferenceSession(self._onnx_forward.SerializeToString())
             self._backward_session = onnxruntime.InferenceSession(self._onnx_backward.SerializeToString())
 
+            if self._device.type == 'cuda':
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:195]: As per documentation, all provders should be used when it is not specified. Maybe we have a bug on `InferenceSession` that we should fix instead:

```
class InferenceSession(Session):
    """
    This is the main class used to run a model.
    """
    def __init__(self, path_or_bytes, sess_options=None, providers=None, provider_options=None):
        """
        :param path_or_bytes: filename or serialized ONNX or ORT format model in a byte string
        :param sess_options: session options
        :param providers: list of providers to use for session. If empty, will use all available providers.
        :param provider_options: list of provider options dict for each provider, in the same order as 'providers'

        The model type will be inferred unless explicitly set in the SessionOptions.
        To explicitly set:
          so = onnxruntime.SessionOptions()
          so.add_session_config_entry('session.load_model_format', 'ONNX') or
          so.add_session_config_entry('session.load_model_format', 'ORT') or

        A file extension of '.ort' will be inferred as an ORT format model.
        All other filenames are assumed to be ONNX format models.
        """
```

Comment (mrry) [reply orttraining/orttraining/python/training/ortmodule.py:195]: What's the bug? The documentation on InferenceSession doesn't constrain what we do in ORTModule.

Either way, we only have training kernels registered in the CPU and CUDA providers, so I'm not sure what the effect would be of opening this up to other providers....

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+                        help='views forward and backward graphs')
+    parser.add_argument('--epochs', type=int, default=10, metavar='N',
+                        help='number of epochs to train (default: 10)')
+    parser.add_argument('--log-level', choices=['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'], default='WARNING',
+                        help='Log level (default: WARNING)')
+
+    # DeepSpeed-related settings
+    parser.add_argument('--local_rank',
+                                                type=int,
+                                                default=-1,
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py:154]: should it be -1 or 0?

Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py]: I took this from the DeepSpeed example code [here](https://github.com/microsoft/DeepSpeedExamples/blob/c3aa269b14ab48df9e745aa0948a7c7d62251197/cifar/cifar10_deepspeed.py#L36). My guess is that it's set to -1 in the example to act as a sentinel if you don't set it, but on reflection it might make more sense to make it a required arg (which I just did in the latest commit).

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
+"""Test for a simple ORTModule using the high-level DeepSpeed API.
+
+To run on the local GPU(s):
+
+```
+$ pip install deepspeed
+$ deepspeed orttraining_test_ortmodule_mnist_deepspeed.py \
```
Comment (spandantiwari) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py:7]: nit: `python orttraining_test_ortmodule....'?

Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py:7]: No, this is using the `deepspeed` launcher binary that is added to your `$PATH` when you install DeepSpeed. It sets various environment variables and flags used in the test.

[the remark this answers is not in the document]
Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py:7]: It doesn't use `DistributedDataParallel` (as far as I can tell), but instead there are explicit calls to `dist.all_reduce()` when you invoke `DeepSpeedEngine.backward(loss)`: https://github.com/microsoft/DeepSpeed/blob/a9a83a6fcfcf654d75017453fbb3a476000180ce/deepspeed/runtime/engine.py#L869

It'll probably make sense to have some simple tests that cover `DistributedDataParallel` as well. I've run a simple model with PyTorch Lightning using its DDP backend, and *that* worked with `DistributedDataParallel`, although we have a few bugs to fix before that's ready to check in.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+    # Reset the total loss for this epoch.
+    total_loss = 0
+
+    for iteration, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        data = data.reshape(data.shape[0], -1).half()
+
+        optimizer.zero_grad()
+        if args.pytorch_only:
```
[the remark this answers is not in the document]
Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py:53]: Agreed... this is copied from Thiago's original test which has the same artefact (and I think wasn't subjected to any code review :)...).

Even though it looks a bit silly, I'm going to keep this as-is to minimize the diff between this file and orttraining_test_ortmodule_mnist_deepspeed.py. We could and should clean up the tests in a subsequent PR.

**Comment by mrry** [comment]
> @mrry Could you add the samples you used to validate both DDP and Zero 1/2?
> 
> Yu can copy & paste the samples in new files if it makes easier. I will work with Baiju to get a proper CI workflow going this week

I added the DeepSpeed DDP+ZeRO test. The PyTorch DDP test that I ran used PyTorch Lightning and requires some more fixes to the device handling logic (on the internal bug tracker), so we'll need to come up with sustainable fixes for those before we can check them in.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py (hunk 6 above)
Comment (spandantiwari) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_mnist_deepspeed.py:7]: OK.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 8]
```diff
@@ -24,14 +24,39 @@
...
+        # could be 'cuda:0', 'cuda:1', or 'cpu'. with cpu, set index=0
+        if device.find(':') == -1:
+            device += ':' + str(torch.cuda.current_device())
+    elif isinstance(device, int):
+        device = 'cuda:' + str(device)
+    elif isinstance(device, torch.device):
+        if device.index is None:
+            device = device.type + ':' + str(torch.cuda.current_device())
+        else:
+            device = device.type + ':' + str(device.index)
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:47]: optional: you could just do str(device).

File: orttraining/orttraining/python/training/ortmodule.py [hunk 9]
```diff
@@ -24,14 +24,39 @@
...
+    elif device is None:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:48]: this can be else instead so we capture all other types as well (although it may never come to that).

File: orttraining/orttraining/python/training/ortmodule.py [hunk 10]
```diff
@@ -98,60 +125,49 @@ def __init__(self, module):
...
-            device = str(device)
-        else:
-            device = 'cuda'
-        if self._device != str(device):
+
+        if device is None and _get_device_str(self._device) != _get_default_device_str('cuda'):
             self._require_export = True
-            self._device = device
+            self._device = torch.device(_get_default_device_str('cuda'))
+        elif _get_device_str(self._device) != _get_device_str(device):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:141]: If the ```device``` is ```None``` and we hit this ```elif``` block, our code would raise an exception: ```'Unsupported device type'```. Is that expected behavior?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 11]
```diff
@@ -185,8 +201,23 @@ def forward(self, *inputs, **kwargs):
...
-            self._backward_session = onnxruntime.InferenceSession(self._onnx_backward.SerializeToString())
+
+            providers = None
+            provider_options = None
+            if self._device.type == 'cuda':
+                # Configure the InferenceSessions to use the specific GPU on which the model is placed.
+                providers = ["CUDAExecutionProvider", "CPUExecutionProvider"]
+                provider_options = [{"device_id": str(self._device.index)}]
+
+                # provider_options = [{}]
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:212]: may we remove dead code?

File: run_ortmodule_mvp_mnist_deepspeed.sh [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
```
Comment (baijumeswani) [inline run_ortmodule_mvp_mnist_deepspeed.sh:1]: how is this script to be leveraged? Is it mostly for dev work? Or is it to be incorporated in ci testing somehow?

File: orttraining/orttraining/python/training/ortmodule.py (hunk 8 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:47]: `str(torch.device('cuda')) != 'cuda:0'`, so we have to create this string manually to pass the device id to ort api

File: orttraining/orttraining/python/training/ortmodule.py (hunk 10 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:141]: Nope. will break this expression in 2

File: run_ortmodule_mvp_mnist_deepspeed.sh (hunk 12 above)
Comment (spandantiwari) [reply run_ortmodule_mvp_mnist_deepspeed.sh:1]: I think we should check it in for now. Once you have the test script ready to run with CI, we should lock this in with a test point.

Comment (thiagocrepaldi) [reply run_ortmodule_mvp_mnist_deepspeed.sh:1]: To serve as a sample, especially for the new comers. It can be deleted once CI is up and running

**Review by baijumeswani — APPROVED** [review]
Looks good.

**Review by thiagocrepaldi — APPROVED** [review]

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]
Waiting for a pipeline to pass

**Review by spandantiwari — APPROVED** [review]
LGTM.

**Review by thiagocrepaldi — APPROVED** [review]
